### ISSUE TITLE
Replace strncpy/vim_strncpy/STRNCPY with strlcpy

### ISSUE THREAD

**Issue by aktau** [neovim/neovim#731]
[strncpy](http://man7.org/linux/man-pages/man3/strcpy.3.html) is [seriously awful](http://stackoverflow.com/a/6987247/558819) (also see these two LWN articles [[1](http://lwn.net/Articles/507319/)], [[2](https://lwn.net/Articles/33812/)]). It actually returns a non-NULL terminated string if the source string is too long. As suggested in that SO post, we should replace it point blank with `strlcpy`. Didn't check if it was available everywhere but we could shim it on the platforms that don't provide it. (yes, that might be an #ifdef, but it'd be very localized and we have to do that for (v)snprintf on windows anyway).

This seems to be a good newcomer issue. **Bonus points for integrating a clint rule that warns on `strncpy` usage.** @philix do we have a plan for marking newcomer issues?

**NOTE**: as far as I can see [strndup](http://man7.org/linux/man-pages/man3/strdup.3.html) does not suffer from this problem, so for allocating-and-copying, there's no problem in the vim codebase.

``` bash
➜  src git:(master) ag -i 'strncpy'
nvim/api/vim.c
122:  strncpy(string, dir.data, dir.size);

nvim/buffer.c
2138:      vim_strncpy(NameBuff, buf_spname(buf), MAXPATHL - 1);
2486:    vim_strncpy(p, buf_spname(curbuf), IOSIZE - (p - buffer) - 1);
2632:        vim_strncpy(buf, (char_u *)_("[No Name]"), SPACE_FOR_FNAME);
2635:        vim_strncpy(buf, p, SPACE_FOR_FNAME);
2667:          vim_strncpy(buf + off, (char_u *)_("help"),
2677:          vim_strncpy(buf + off, p, (size_t)(SPACE_FOR_DIR - off));
2680:          vim_strncpy(buf + off, (char_u *)"...",
3084:        vim_strncpy(NameBuff, buf_spname(wp->w_buffer), MAXPATHL - 1);
3528:    vim_strncpy(buf, (char_u *)(above == 0 ? _("All") : _("Bot")),
3531:    vim_strncpy(buf, (char_u *)_("Top"), (size_t)(buflen - 1));

nvim/edit.c
3802:               * compl_length, so the next STRNCPY always
3804:              STRNCPY(IObuff, ptr, len);
3827:                STRNCPY(IObuff + len, ptr, tmp_ptr - ptr);

nvim/eval.c
6360:    vim_strncpy(buf, key, len);
19055:  STRNCPY(IObuff + 8, s, IOSIZE - 8);

nvim/ex_cmds.c
322:    STRNCPY(sortbuf1, ml_get(l1.lnum) + l1.start_col_nr,
325:    STRNCPY(sortbuf2, ml_get(l2.lnum) + l2.start_col_nr,

nvim/ex_getln.c
3442:      vim_strncpy(retval, fname, len);
3974:    vim_strncpy(buf, s, l);
3977:    vim_strncpy(buf + l, pat, MAXPATHL - 1 - l);

nvim/file_search.c
321:      vim_strncpy(ff_expand_buffer, rel_fname, len);
1134:    vim_strncpy(ff_expand_buffer, fname, MAXPATHL - 1);

nvim/api/helpers.c
78:    strncpy(err->msg, msg, sizeof(err->msg));

nvim/api/helpers.h
11:    strncpy(err->msg, message, sizeof(err->msg));  \

nvim/fold.c
1616:      vim_strncpy(newline + line_len, marker, markerlen);
1619:      STRNCPY(newline + line_len + (p - cms), marker, markerlen);
1685:        STRNCPY(newline, line, p - line);

nvim/getchar.c
284:    vim_strncpy(buf->bh_curr->b_str + len, s, (size_t)slen);
293:    vim_strncpy(p->b_str, s, (size_t)slen);

nvim/hardcopy.c
1600:  vim_strncpy(resource->name, (char_u *)name, 63);
1819:      vim_strncpy(resource->title, dsc_line.string, dsc_line.len);
1826:      vim_strncpy(resource->version, dsc_line.string, dsc_line.len);
2231:        vim_strncpy((char_u *)prt_cmap,
2241:      vim_strncpy((char_u *)prt_cmap, p_pmcs, sizeof(prt_cmap) - 3);

nvim/if_cscope.c
334:  vim_strncpy(buf, (char_u *)p, size - 1);
2093:    vim_strncpy(csdir, (char_u *)csinfo[i].fname,

nvim/fileio.c
2374:  vim_strncpy(header, (char_u *)crypt_magic[use_crypt_method],

nvim/mark.c
480:      vim_strncpy(NameBuff + len, fm->fname + 2, MAXPATHL - len - 1);
482:      vim_strncpy(NameBuff, fm->fname, MAXPATHL - 1);

nvim/memline.c
342:  STRNCPY(b0p->b0_version, "VIM ", 4);
343:  STRNCPY(b0p->b0_version + 4, Version, 6);
856:        vim_strncpy(b0p->b0_fname, buf->b_ffname,
1148:    vim_strncpy(NameBuff, buf_spname(curbuf), MAXPATHL - 1);
3285:  vim_strncpy(tmp, fname, MAXPATHL - 1);

nvim/message.c
1900:    vim_strncpy(mp->sb_text, *sb_str, s - *sb_str);

nvim/misc1.c
512:      vim_strncpy(leader, saved_line, lead_len);
2600:        vim_strncpy(msg_buf, (char_u *)_("1 more line"),
2603:        vim_strncpy(msg_buf, (char_u *)_("1 line less"),
3152:    vim_strncpy(name, (char_u *)envname, ENVNAMELEN - 1);

nvim/misc2.c
824:  vim_strncpy(dir, fname, MAXPATHL - 1);

nvim/ops.c
2250:      vim_strncpy(ins_text, firstline + bd.textcol, (size_t)ins_len);

nvim/option.c
3284:      vim_strncpy(IObuff, (char_u *)_(errmsg), IOSIZE - 1);
7632:      vim_strncpy(NameBuff, varp, MAXPATHL - 1);

nvim/os/env.c
71:    strncpy(hostname, vutsname.nodename, len - 1);

nvim/os/fs.c
32:    vim_strncpy(buf, (char_u *)uv_strerror(errno), len - 1);
105:    vim_strncpy(buf, (char_u *) path, e - path);

nvim/os/server.c
81:  strncpy(addr, endpoint, sizeof(addr));
107:  strncpy(ip, addr, addr_len);
179:  strncpy(addr, endpoint, sizeof(addr));

nvim/os/users.c
61:    vim_strncpy((char_u *)s, (char_u *)pw->pw_name, len - 1);

nvim/menu.c
910:      vim_strncpy(path_name, arg, after_dot - arg - 1);
1026:        vim_strncpy(tbuffer, menu->en_dname, TBUFFER_LEN - 2);
1028:        vim_strncpy(tbuffer, menu->dname,  TBUFFER_LEN - 2);

nvim/path.c
410:      STRNCPY(p, path_end, len);
1532:    vim_strncpy(buf, fname, len - 1);
1960:      STRNCPY(relative_directory, fname, p-fname);

nvim/os_unix.c
414:          vim_strncpy(newname, name, MAXPATHL);
415:          vim_strncpy(newname + (tail - name), (char_u *)dp->d_name,

nvim/quickfix.c
519:            vim_strncpy(IObuff, p_str, CMDBUFFSIZE - 2);
521:            vim_strncpy(IObuff, p_str, len);
536:          vim_strncpy(IObuff, p_li->li_tv.vval.v_string, len);
544:        vim_strncpy(IObuff, ml_get_buf(buf, buflnum++, FALSE),
644:          vim_strncpy(errmsg, regmatch.startp[i], len);

nvim/regexp.c
6853:          vim_strncpy(retval, s, len);
6875:          STRNCPY(retval + len, reg_getline_submatch(lnum),

nvim/ex_docmd.c
3577:        STRNCPY(ptr, program, i);
4245:          vim_strncpy(buff,
6034:        vim_strncpy(IObuff, buf_spname(wp->w_buffer), IOSIZE - 1);

nvim/spell.c
1472:              vim_strncpy(fword, ptr, endlen[endidxcnt]);
2195:      vim_strncpy(buf +  n, p, maxlen - 1 - n);
2992:    vim_strncpy(buf, word, len);
3288:    vim_strncpy(syl->sy_chars, s, l);
3699:        vim_strncpy(region_cp, p + 1, 2);
3821:        vim_strncpy(lang, path_tail(spf_name), MAXWLEN);
3942:          vim_strncpy(bp + n, p, l);
4847:        vim_strncpy(key, items[1], AH_KEY_LEN - 1);
5420:        vim_strncpy(key, prevp, p - prevp);
5809:      vim_strncpy(key, prevp, p - prevp);
5840:      vim_strncpy(key, prevp, p - prevp);
5935:                vim_strncpy(newword, ae->ae_add, MAXWLEN - 1);
5949:              vim_strncpy(newword, word, MAXWLEN - 1);
6031:              vim_strncpy(pfx_pfxlist, use_pfxlist, use_pfxlen);
7386:  vim_strncpy(fname, wfname, MAXPATHL - 1);
7811:      vim_strncpy(wfname, fnames[0], MAXPATHL - 1);
8173:        vim_strncpy(buf, curbuf->b_s.b_p_spl,
8182:          vim_strncpy(buf, curbuf->b_s.b_p_spl,
8733:      vim_strncpy(wcopy, stp->st_word, MAXWLEN);
8735:        vim_strncpy(wcopy + stp->st_wordlen,
9024:  vim_strncpy(su->su_badword, su->su_badptr, su->su_badlen);
9470:  vim_strncpy(wcopy + l, p, MAXWLEN - l - 1);
9837:          vim_strncpy(preword + sp->ts_prewordlen,
10975:          STRNCPY(kword + kwordlen[depth], fword + fwordidx[depth],
10979:          STRNCPY(kword + kwordlen[depth], uword + uwordidx[depth],
11184:    vim_strncpy(goodword + stp->st_wordlen,
11774:    vim_strncpy(longword, stp[i].st_word, MAXWLEN);
11776:    vim_strncpy(longword + len, su->su_badptr + stp[i].st_orglen,
12047:    vim_strncpy(word, s, MAXWLEN - 1);
13509:            vim_strncpy(prefix + depth, word, MAXWLEN - depth - 1);
13524:              vim_strncpy(prefix + depth, word_up,

nvim/strings.c
70:  STRNCPY(p, string, len);
342: * Like strncpy(), but always terminate the result with one NUL.
345:void vim_strncpy(char_u *to, char_u *from, size_t len)
347:  STRNCPY(to, from, len);

nvim/strings.h
17:void vim_strncpy(char_u *to, char_u *from, size_t len);

nvim/syntax.c
3016:  vim_strncpy(keyword, kwp, kwlen);
5247:      vim_strncpy(name + 1, p, end - p);
7380:  vim_strncpy(name_u, name, 199);

nvim/regexp_nfa.c
2365:    STRNCPY(save, &p[last], 2);
2366:    STRNCPY(&p[last], "+-", 2);
2368:    STRNCPY(&p[last], save, 2);

nvim/tag.c
702:          vim_strncpy(tag_name, tagp.tagname, len);
709:          vim_strncpy(fname, p, MAXPATHL);
1858:                  vim_strncpy(p, tagp.command + 2, len);
1868:                vim_strncpy(p, tagp.tagname, len);
2092:      vim_strncpy(buf, ((char_u **)(tag_fnames.ga_data))[
2682:    vim_strncpy(retval + (p - tag_fname), fname,
2842:    vim_strncpy(buf, start, len);

nvim/version.c
865:    vim_strncpy(modby, (char_u *)_("Modified by "), MODBY_LEN - 1);
867:    vim_strncpy(modby + l, (char_u *)MODIFIED_BY, MODBY_LEN - l - 1);

nvim/screen.c
2043:        STRNCPY(current_ScreenLine, text, len);
2045:        STRNCPY(current_ScreenLine + col, text, len);
4771:          STRNCPY(buf + len, s, l);
5161:  vim_strncpy(buf, p, sizeof(buf) - 1);
7706:    vim_strncpy(NameBuff, buf_spname(buf), MAXPATHL - 1);

nvim/vim.h
978:#define STRNCPY(d, s, n)    strncpy((char *)(d), (char *)(s), (size_t)(n))

nvim/search.c
4343:          /* IOSIZE > compl_length, so the STRNCPY works */
4344:          STRNCPY(IObuff, aux, i);
4378:            STRNCPY(IObuff + i, aux, p - aux);
```

**Comment by Breakster** [comment]
Hello vimmers,

This project is really interesting. I would like to contribute, so I was looking at the entry-level issues to see if there is something I could pick up.

Looking into this issue, I noticed that `strncpy` and `STRNCPY` are still being used several times in the code. Maybe they can be replaced by `STRLCPY` instead. This a macro wrapper around `xstrlcpy`. This would avoid writing an `#ifdef` for platforms that don't support `strlcpy`, because `xstrlcpy` is already being used.

I would like to continue with this issue, if it is deemed useful.

**Comment by aktau** [comment]
@Breakster you're welcome to. Do keep in mind that because of some issues we encountered during the automatic conversion, this no longer seems to be an entry-level issue.

I've been contemplating some techniques to deduce where (neo)vim is relying on `strncpy`'s NUL-blasting behaviour. One of them is to use clang/llvm's adress/memory sanitizer infrastructure to manually "poison" these NUL-blasted bytes. Such that the program aborts of nvim tries to read it. Have a look here: https://code.google.com/p/address-sanitizer/wiki/ManualPoisoning

**Comment by EJahren** [comment]
I have been looking into this somewhat, but I wonder if there is a better solution to go away from c strings all together. There is a lot of use of snprintf/vim_snprintf/sprintf/garray/strcat/strcpy/strncpy/vim_strsave with varying degrees of safety.

I haven't seen a single use of `snprintf` that checks that printing was successful, and unsuccessful `snprintf`s results in non-null terminated strings.

There are several cases where the use of c strings can easily be changed to use glibs  `GSring` instead. This means that one is at least guaranteed safe printfs and concats. As a bonus, the `GString` keeps the length of the string in the header so it might actually be a bit faster now and then.

Granted, this means that neovim will depend upon glib.

**Comment by splinterofchaos** [comment]
Personally, I like glib, it's widely available, and it provides many OS abstractions where we currently use libuv, which is really more of an event loop abstraction. As for porting large portions of the codebase to glib for string management... I think `xstrlcpy` is quite sufficient. I also think dependencies should be minimal.

I'm not opposed to the idea, just have some concerns.

**Comment by philix** [comment]
Much of the useful glib functionality (useful to Neovim) is already covered by `libuv` (e.g. platform-independent threads, filesystem) or implemented by us (e.g. memory allocation functions). Glib would be very redundant. C strings are good:tm: and Pascal strings (struct with buffer and length) are used where they seem more appropriate.

@aktau proposed [sds](https://github/antirez/sds) https://github.com/neovim/neovim/issues/859#issuecomment-46429639 which is IMO a better way to mix C strings and length-aware strings.

In fact, sds was proposed since the very early issue #210.

**Comment by EJahren** [comment]
sds seems lik a reasonable choice, although since the string type is only defined by `typedef char *sds`
you wont get a compiler warning if you use a regular string as a sds string. Additionally coverety might
get a lot of false negatives and false positives on determining whether a given string is safe to use as a sds string. Is it really that bad to have to grab the \0 terminated buffer from a struct when you need the c string?

**Comment by flyingmutant** [comment]
FYI, there is a new proposed `strscpy` primitive for Linux, intended to replace most uses of `strlcpy` and `strncpy`.

- [initial patch (+ motivation)](https://lkml.org/lkml/2015/4/30/582)
- [current implementation + documentation](https://git.kernel.org/cgit/linux/kernel/git/cmetcalf/linux-tile.git/commit/?h=strscpy&id=2080420588c843b85f6377d6b33b9bf13d61a4e3)

**Comment by aktau** [comment]
> FYI, there is a new proposed strscpy primitive for Linux, intended to replace most uses of strlcpy and strncpy.

Yep, I also mentioned it in a (now closed) old PR of mine: #869.

**Comment by splinterofchaos** [comment]
> Much of the useful glib functionality (useful to Neovim) is already covered by libuv (e.g. platform-independent threads, filesystem) 

glib has many other things to offer as well, like [character set conversion](https://developer.gnome.org/glib/2.44/glib-Character-Set-Conversion.html), [unicode support](https://developer.gnome.org/glib/2.44/glib-Unicode-Manipulation.html) (which might be insufficient for our needs, though), [timers](https://developer.gnome.org/glib/2.44/glib-Timers.html) (as disscussed in the past, libuv timers may not be optimal, although that might have been solved by now), and the functions labelled under [misc](https://developer.gnome.org/glib/2.44/glib-Miscellaneous-Utility-Functions.html) could easily repleace parts of `path.c`, `os/env.c`, and probably be less buggy. Also, the [string utilities](https://developer.gnome.org/glib/2.44/glib-String-Utility-Functions.html), working on C-strings, look pretty decent. They even have OS-independent [mmaped files](https://developer.gnome.org/glib/2.44/glib-File-Utilities.html#g-mapped-file-new)! 

It also implements [generic data types](https://developer.gnome.org/glib/2.44/glib-data-types.html) ranging from simple linked lists and and dynamic arrays (`garray_T`/`DynamicBuffer`/`klist`) to hash tables, trees, and variant objects.

Libuv abstracts the event loop and IO, but I wouldn't really call it an OS-abstraction. Calling uv functions with `NULL` as the callback, but still needing to pass `uv_default_loop()`, as we do often in `os/fs.c` and other places, makes me feel like we're using it wrong.
